File: toys/mmap_recv.c
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/mman.h>
#include <unistd.h>
#include <string.h>

int main() {
	int threads = 16;
	size_t size = threads * 2e5;
	int fd = open("/tmp/testmmap", O_RDWR);
	volatile char *buf = (char*)mmap(NULL, 4096 + size, PROT_READ|PROT_WRITE, MAP_SHARED, fd, 0);
	char *dst = (char*)malloc(size);
	for (size_t i = 0; i < size; i++) {
		dst[i] = 0xff & i;
	}
	#pragma omp parallel for
	for (size_t i = 0; i < threads; i++) {
		size_t off = i * size/threads;
		size_t k = 0;
		size_t res = 0;
		for (int j = 0; j < 1e5; j++) {
			while (buf[i] == 1) {
				k = (k + 1) % (size / threads);
				res ^= dst[off+k] ^ j;
			}
			// printf("recv %zd, %d\n", i, j);
			// memcpy(buf+4096, src, size);
			memcpy(dst+off, buf+(off+4096), size/threads);
			buf[i] = 1;
		}
		printf("%zd: %zd\n", i, res);
	}
	ftruncate(fd, 0);
	return 0;
}

File: toys/mmap_recv.js
const mmap = require('mmap.js');
const fs = require('fs');

const threads = 4;
const msgSize = 2048*2048;
const size = threads * msgSize;
const fd = fs.openSync("/tmp/testmmap", 'r+');
const buf = mmap.alloc(4096 + size, mmap.PROT_READ | mmap.PROT_WRITE, mmap.MAP_SHARED, fd, 0);

const slices = [];
for (var i=0 ; i < threads; i++) {
    slices.push(new Uint8Array(buf.buffer, 4096 + i * msgSize, msgSize));
}

const dst = Buffer.alloc(size);

for (var j = 0; j < 1e2; j++) {
    for (var i = 0; i < threads; i++) {
        while (buf[i] === 1) {
        }
        dst.set(slices[i], i * msgSize);
        buf[i] = 1;
    }
}

fs.ftruncateSync(fd, 0);
fs.closeSync(fd);

for (var i = 0; i < threads; i++) {
    for (var j=0; j < msgSize; j++) {
        if (dst[i*msgSize + j] !== i) {
            console.log("mismatch at", i*msgSize + j, ":", dst[i*msgSize + j], "!==", i);
            process.exit();
        }
    }
}
console.log("All a-ok!");
